### PR TITLE
Decouple scope storage between Datadog.Trace versions

### PR DESCRIPTION
**Pull request by colin-higgins**
Pull all serialized classes and values into independent versioned assembly

@DataDog/apm-dotnet

### CODE REVIEW

File: src/Datadog.Trace/Tags.cs [hunk 1]
```diff
@@ -7,28 +7,6 @@ namespace Datadog.Trace
...
-        /// <summary>
-        /// The name of the method that was instrumented to generate the span.
-        /// </summary>
-        public const string InstrumentedMethod = "instrumented.method";
-
-        /// <summary>
-        /// The kind of span (e.g. client, server). Not to be confused with <see cref="Span.Type"/>.
-        /// </summary>
-        /// <seealso cref="SpanKinds"/>
-        public const string SpanKind = "span.kind";
```
Comment (colin-higgins) [inline src/Datadog.Trace/Tags.cs]: I'm not against having proxies in here that point to CoreTags in order to reduce the breaking changes, but I don't really *like* it.

File: src/Datadog.Trace/Vendors/Serilog/Core/Logger.cs [hunk 2]
```diff
@@ -33,7 +33,7 @@ namespace Datadog.Trace.Vendors.Serilog.Core
     /// be disposed to flush any events buffered within it. Most application
     /// code should depend on <see cref="ILogger"/>, not this class.
     /// </summary>
-    internal sealed class Logger : ILogger, ILogEventSink, IDisposable
+    internal sealed class Logger : ILogger, ILogEventSink, IDisposable, ICoreLogger
```
Comment (colin-higgins) [inline src/Datadog.Trace/Vendors/Serilog/Core/Logger.cs]: Adding ICoreLogger here lets us pull this into Core as a strategy

File: src/Datadog.Trace.Core/Logging/LogEventLevel.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+namespace Datadog.Trace
```
Comment (colin-higgins) [inline src/Datadog.Trace.Core/Logging/LogEventLevel.cs]: This seemed like the easiest way to allow log level checks, it will require an update to the vendors tool, to ignore that file on update.

File: src/Datadog.Trace.Core/Logging/CoreLogStrategy.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
+using System;
+
+namespace Datadog.Trace
+{
+    internal static class CoreLogStrategy
+    {
+        private static Func<Type, ICoreLogger> _strategy = null;
+
+        public static void SetStrategy(Func<Type, ICoreLogger> strategy)
```
Comment (colin-higgins) [inline src/Datadog.Trace.Core/Logging/CoreLogStrategy.cs]: To be called when the Datadog.Trace logger is ready,

File: src/Datadog.Trace.Core/Datadog.Trace.Core.csproj [hunk 5]
```diff
@@ -0,0 +1,10 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <!-- NuGet -->
+    <Version>0.1.0-prerelease</Version>
+    <Title>Context Storage for Datadog APM</Title>
+    <Description>Context storage and management for Datadog .NET Tracer library.</Description>
+  </PropertyGroup>
+
+</Project>
```
Comment (lucaspimentel) [inline src/Datadog.Trace.Core/Datadog.Trace.Core.csproj]: We may want to add `<RootNamespace>Datadog.Trace</RootNamespace>` so IDEs don't keep suggesting we change all the namespaces in this project to `Datadog.Trace.Core`.

File: src/Datadog.Trace.Core/Scope.cs [hunk 6]
```diff
@@ -19,6 +22,11 @@ internal Scope(Scope parent, Span span, IScopeManager scopeManager, bool finishO
             Span = span;
             _scopeManager = scopeManager;
             _finishOnClose = finishOnClose;
+
+            if (_log == null)
```
Comment (lucaspimentel) [inline src/Datadog.Trace.Core/Scope.cs]: Won't `_log` always be null here since this is the ctor? Looks like we can make `_log` readonly, too.

File: src/Datadog.Trace/Logging/DatadogLogging.cs [hunk 7]
```diff
@@ -133,6 +132,10 @@ static DatadogLogging()
                 Initialized = true;
                 // nothing else to do here
             }
+            finally
+            {
+                CoreLogStrategy.SetStrategy(GetSharedLogger);
```
Comment (colin-higgins) [inline src/Datadog.Trace/Logging/DatadogLogging.cs]: This can probably be the first thing that happens actually, as it will be lazily called for the classes that need it.

File: src/Datadog.Trace.Core/Scope.cs (hunk 6 above)
Comment (colin-higgins) [reply src/Datadog.Trace.Core/Scope.cs]: readonly can be assigned in a constructor, but static readonly can only be assigned in static constructor or variable initializer

Comment (colin-higgins) [reply src/Datadog.Trace.Core/Scope.cs]: Also, it's a static to prevent re-assignment every time. It won't hurt to make it per instance and assign every time though, I think.

Comment (colin-higgins) [reply src/Datadog.Trace.Core/Scope.cs]: Because what is returned is a single SharedLogger instance anyways.

Comment (lucaspimentel) [reply src/Datadog.Trace.Core/Scope.cs]: Never mind. I missed that is was `static` 😉

File: src/Datadog.Trace/Tags.cs (hunk 1 above)
Comment (lucaspimentel) [reply src/Datadog.Trace/Tags.cs]: Agreed. Why not move the entire `Tags` class to the new assembly?

Comment (colin-higgins) [reply src/Datadog.Trace/Tags.cs]: Because we add to it based on new integrations. This seemed like the best way to not pollute core with non-core things.

Comment (zacharycmontoya) [reply src/Datadog.Trace/Tags.cs]: These are ultimately just strings, so I don't think they need to be in the Core assembly. The main concern with that assembly should be types that have a well-formed shape

Comment (colin-higgins) [reply src/Datadog.Trace/Tags.cs]: @zacharycmontoya A lot of TraceContext and SpanContext were pretty tangled with many tags but not others.

File: tools/UpdateVendors/Program.cs [hunk 8]
```diff
@@ -17,6 +17,11 @@ public class Program
 //------------------------------------------------------------------------------
 ";
 
+        private static readonly List<string> FilesToSkip = new List<string>()
+        {
+            @"\Serilog\Events\LogEventLevel.cs",
```
Comment (zacharycmontoya) [inline tools/UpdateVendors/Program.cs:22]: Is this new list being consumed yet?

Comment (colin-higgins) [reply tools/UpdateVendors/Program.cs:22]: Nope, another merge problem, thanks.

File: src/Datadog.Trace/Tags.cs [hunk 9]
```diff
@@ -152,19 +152,7 @@ public static class Tags
         /// <summary>
         /// The sampling priority for the entire trace.
         /// </summary>
-        public const string SamplingPriority = "sampling.priority";
-
-        /// <summary>
-        /// Obsolete. Use <see cref="ManualKeep"/>.
-        /// </summary>
-        [Obsolete("This field will be removed in futures versions of this library. Use ManualKeep instead.")]
-        public const string ForceKeep = "force.keep";
```
Comment (zacharycmontoya) [inline src/Datadog.Trace/Tags.cs:161]: Can you keep these two constants in the `Datadog.Trace` assembly so there's no breaking change? Or do you also need the tags in the `Datadog.Trace.Core` assembly?

File: src/Datadog.Trace/Tags.cs [hunk 10]
```diff
@@ -10,7 +10,7 @@ public static class Tags
         /// <summary>
         /// The environment of the profiled service.
         /// </summary>
-        public const string Env = "env";
+        public const string Env = CoreTags.Env;
 
         /// <summary>
         /// The version of the profiled service.
```
Comment (zacharycmontoya) [inline src/Datadog.Trace/Tags.cs:16]: Can you also add this `"version"` tag to the CoreTags? Now that we're targeting the service/env/version trio

File: src/Datadog.Trace/Tags.cs (hunk 1 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace/Tags.cs:35]: With fresh eyes, I'm pretty happy with the way the changes look which includes several proxies for key things and leaves the rest in the `Datadog.Trace` assembly which can be added as we add integrations

File: src/Datadog.Trace/Tags.cs (hunk 9 above)
Comment (colin-higgins) [reply src/Datadog.Trace/Tags.cs:161]: There was originally a need to keep them in Core, but with the way I abstracted away the Span storage in the most recent changes, that's not the case anymore.

**Comment by zacharycmontoya** [comment]
One thing I just realized, in our Docs we still offer guidance for manual tracing to use `Tracer.Instance`, where Tracer will be provided by the `Datadog.Trace` assembly so users, if they're still mismatching NuGet version and Profiler version would get different instances and their Traces will still be broken. Is that a goal for this PR to solve?

File: src/Datadog.Trace/Tracer.cs [hunk 11]
```diff
@@ -109,7 +118,10 @@ internal Tracer(TracerSettings settings, IAgentWriter agentWriter, ISampler samp
             // fall back to default implementations of each dependency if not provided
             _agentWriter = agentWriter ?? new AgentWriter(new Api(Settings.AgentUri, delegatingHandler: null, Statsd), Statsd);
 
-            _scopeManager = scopeManager ?? new AsyncLocalScopeManager();
+            // Central Datadog.Trace version independent scope storage
+            CoreStorage.ScopeManager = scopeManager ?? new AsyncLocalScopeManager();
```
Comment (zacharycmontoya) [inline src/Datadog.Trace/Tracer.cs:122]: In the use case where our automatic instrumentation gets the Tracer via `Tracer.Instance` and then the user code gets the Tracer (different version) via `Tracer.Instance`, I think the code as-is will replace the `CoreStorage.ScopeManager` with a new scope manager when the second call is made. Here's my suggestion:

```
_scopeManager = scopeManager ?? CoreStorage.ScopeManager ?? new AsyncLocalScopeManager();
CoreStorage.ScopeManager = _scopeManager;
```

Comment (colin-higgins) [reply src/Datadog.Trace/Tracer.cs:122]: Should we only ever set `CoreStorage.ScopeManager` once and have an override method?

File: src/Datadog.Trace.Core/Datadog.Trace.Core.csproj [hunk 12]
```diff
@@ -0,0 +1,13 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+  
+    <!-- NuGet -->
+    <Version>1.1.0</Version>
```
Comment (lucaspimentel) [inline src/Datadog.Trace.Core/Datadog.Trace.Core.csproj:6]: Leaving a comment here as a reminder to change this back to 1.0.0 before merging.

File: src/Datadog.Trace.Core/TraceContext.cs [hunk 13]
```diff
@@ -1,26 +1,29 @@
...
 
+        private readonly Action<Span[]> _submit = null;
+        private readonly Func<Span, SamplingPriority> _getSamplingPriority;
+
         private int _openSpans;
         private SamplingPriority? _samplingPriority;
         private bool _samplingPriorityLocked;
 
-        public TraceContext(IDatadogTracer tracer)
+        public TraceContext(Action<Span[]> submit, Func<Span, SamplingPriority> getSamplingPriority)
```
Comment (lucaspimentel) [inline src/Datadog.Trace.Core/TraceContext.cs:23]: Why are we removing the `IDatadogTracer` parameter here in favor of delegates?

File: src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs [hunk 14]
```diff
@@ -147,28 +147,28 @@ private void OnHostingHttpRequestInStart(object arg)
             }
             else
             {
-                HttpRequest request = httpContext.Request;
-                string host = request.Host.Value;
-                string httpMethod = request.Method?.ToUpperInvariant() ?? "UNKNOWN";
-                string url = GetUrl(request);
+                var request = httpContext.Request;
```
Comment (lucaspimentel) [inline src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs:150]: Please avoid unrelated code formatting and style changes.

File: samples/Samples.AspNetCoreMvc31/Properties/launchSettings.json [hunk 15]
```diff
@@ -13,11 +13,9 @@
       "launchBrowser": false,
       "environmentVariables": {
         "ASPNETCORE_ENVIRONMENT": "Development",
-
```
Comment (lucaspimentel) [inline samples/Samples.AspNetCoreMvc31/Properties/launchSettings.json:16]: Please avoid unrelated code formatting and style changes.

File: src/Datadog.Trace.Core/TraceContext.cs (hunk 13 above)
Comment (colin-higgins) [reply src/Datadog.Trace.Core/TraceContext.cs:23]: I did this in favor of moving `IDatadogTracer` to `Datadog.Trace.Core` as it meant fewer code changes and fewer reasons to change in the future.

Happy to accomplish this however, but this felt like a reasonable solution to me, as the minimum external requirements are: (1) How to submit (2) How to get sampling rate

Comment (colin-higgins) [reply src/Datadog.Trace.Core/TraceContext.cs:23]: Perhaps using composition on `IDatadogTracer` by having a core interface that implements those two requirements would be cleaner....
`ICoreDatadogTracer`

Comment (colin-higgins) [reply src/Datadog.Trace.Core/TraceContext.cs:23]: I think another reason I originally did this is because these methods came from two separate interfaces.

Comment (colin-higgins) [reply src/Datadog.Trace.Core/TraceContext.cs:23]: Yeah, `IDatadogTracer` lives in core, but not the sampler.

File: src/Datadog.Trace/Tracer.cs (hunk 11 above)
Comment (lucaspimentel) [reply src/Datadog.Trace/Tracer.cs:122]: Do we even need to replace `_scopeManager` with the static `CoreStorage.ScopeManager`? What does that get us?

File: src/Datadog.Trace/SpanImplementation.cs [hunk 16]
```diff
@@ -15,7 +13,7 @@ namespace Datadog.Trace
     /// tracks the duration of an operation as well as associated metadata in
     /// the form of a resource name, a service name, and user defined tags.
     /// </summary>
-    public class Span : IDisposable, ISpan
+    public class SpanImplementation : Span
```
Comment (lucaspimentel) [inline src/Datadog.Trace/SpanImplementation.cs:16]: Wouldn't there be less code changes overall if this was something like `Span : SpanBase`?
Also, is the new abstract `Span` used anywhere, or should we revert this to `Span : IDisposable, ISpan`

File: src/Datadog.Trace/Tracer.cs (hunk 11 above)
Comment (colin-higgins) [reply src/Datadog.Trace/Tracer.cs:122]: We need to store the ScopeManager ref in the shared `Datadog.Trace.Core` assembly between disparate `Datadog.Trace` versions in order to have manual/automatic traces properly nest.

File: src/Datadog.Trace/Tracer.cs [hunk 17]
```diff
@@ -309,20 +316,20 @@ public Span StartSpan(string operationName, ISpanContext parent = null, string s
             if (parent is SpanContext parentSpanContext)
             {
                 traceContext = parentSpanContext.TraceContext ??
-                               new TraceContext(this)
+                               new TraceContext(_agentWriter.WriteTrace, Sampler.GetSamplingPriority)
```
Comment (lucaspimentel) [inline src/Datadog.Trace/Tracer.cs:319]: This code is now allocating two delegates for every trace. Consider caching the delegates and reusing them.

File: src/Datadog.Trace/SpanImplementation.cs (hunk 16 above)
Comment (colin-higgins) [reply src/Datadog.Trace/SpanImplementation.cs:16]: The abstract Span is used everywhere the original Span was, in both `Datadog.Trace.Core` and `Datadog.Trace`
I did this partially to avoid a slew of renames and fighting with what properties/methods to expose per assembly.
I can look again, but there were problems with the interface last time I tried, I am not recalling specifically what.

File: src/Datadog.Trace.Core/Logging/CoreLogging.cs [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+        public static void SetStrategy(Func<Type, ICoreLogger> strategy)
+        {
+            // First come, first set
+            if (_strategy == null)
+            {
+                _strategy = strategy;
+            }
+            else
+            {
+                // TODO: log that we have loaded multiple versions?
```
Comment (colin-higgins) [inline src/Datadog.Trace.Core/Logging/CoreLogging.cs:18]: Thoughts?

File: src/Datadog.Trace.Core/TraceContext.cs (hunk 13 above)
Comment (colin-higgins) [reply src/Datadog.Trace.Core/TraceContext.cs]: I added a new strategy interface.

File: src/Datadog.Trace/Vendors/Serilog/Core/Logger.cs (hunk 2 above)
Comment (lucaspimentel) [reply src/Datadog.Trace/Vendors/Serilog/Core/Logger.cs:36]: Every time we modify vendor code it makes me wonder if we're doing this right. Did we consider, for example, writing a light wrapper around `Logger` that implements `ICoreLogger`?

Comment (colin-higgins) [reply src/Datadog.Trace/Vendors/Serilog/Core/Logger.cs:36]: I had weighed doing that (with the extra class and needing to implement the constructor passthroughs) and opted for this strategy as changing less and more lightweight.
If you feel strongly either way, I'm okay with creating a wrapper.
